documentation/man/man3/Dealer.3
.TH "Dealer" 3 "Sun Nov 20 2016" "Version 1.0.0.0" "Aeda-Casino" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Dealer \- A dealer\&. The dealer is responsible to deal cards to the players and to himself\&. He is the one that holds that deck and reshufles it\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Dealer\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBDealer\fP ()"
.br
.RI "Default constructor\&. Creates a default deck of 52 cards\&. "
.ti -1c
.RI "\fBDealer\fP (unsigned int ID)"
.br
.RI "Constructor\&. Creates a default deck of 52 cards\&. "
.ti -1c
.RI "\fBDealer\fP (vector< \fBCard\fP > newDeck)"
.br
.RI "Constructor\&. "
.ti -1c
.RI "void \fBshuffleDeck\fP ()"
.br
.RI "Shuffle deck\&. "
.ti -1c
.RI "\fBCard\fP \fBdiscard\fP (vector< \fBPlayer\fP *> players)"
.br
.RI "Removes a \fBCard\fP from the end of the deck and returns it\&. Additionaly pushes the \fBCard\fP to the deckDiscarded vector\&. "
.ti -1c
.RI "\fBCard\fP & \fBhit\fP (vector< \fBPlayer\fP *> players)"
.br
.RI "Places a \fBCard\fP in the \fBDealer\fP hand and updates the handscore\&. "
.ti -1c
.RI "unsigned int \fBsetHandScore\fP ()"
.br
.RI "Updates dealer's handscore\&. "
.ti -1c
.RI "vector< \fBCard\fP > & \fBgetDeck\fP ()"
.br
.RI "Gets the deck\&. "
.ti -1c
.RI "vector< \fBCard\fP > & \fBgetHand\fP ()"
.br
.RI "Gets the hand\&. "
.ti -1c
.RI "void \fBaddCardsToDeck\fP (vector< \fBCard\fP > &cards)"
.br
.RI "Adds cards to the deck\&. "
.ti -1c
.RI "vector< \fBCard\fP > & \fBgetDiscardedDeck\fP ()"
.br
.RI "Gets discarded deck\&. "
.ti -1c
.RI "unsigned int \fBgetHandScore\fP () const"
.br
.RI "Gets handscore\&. "
.ti -1c
.RI "void \fBclearDiscardedDeck\fP ()"
.br
.RI "Clears the discarded deck\&. "
.ti -1c
.RI "void \fBclearHand\fP ()"
.br
.RI "Clears the hand\&. "
.ti -1c
.RI "bool \fBhasCardInHand\fP (\fBCard\fP &c)"
.br
.RI "Query if dealer has card in hand\&. "
.ti -1c
.RI "string \fBplay\fP (\fBTable\fP &table)"
.br
.RI "Calculates dealer next move\&. \fBDealer\fP will stand if handscore >= 17\&. "
.ti -1c
.RI "unsigned int \fBgetID\fP () const"
.br
.RI "Gets the \fBDealer\fP's identifier\&. "
.ti -1c
.RI "void \fBsetTable\fP (int tableID)"
.br
.RI "Associates a table to this dealer\&. "
.ti -1c
.RI "int \fBgetTableOn\fP () const"
.br
.RI "Gets identifier of the table that the dealer is on\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBsetNextID\fP (unsigned int ID)"
.br
.RI "Sets next identifier\&. "
.ti -1c
.RI "static unsigned int \fBgetNextID\fP ()"
.br
.RI "Gets the next identifier\&. "
.in -1c
.SH "Detailed Description"
.PP 
A dealer\&. The dealer is responsible to deal cards to the players and to himself\&. He is the one that holds that deck and reshufles it\&. 


.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016 
.RE
.PP

.PP
Definition at line 30 of file Dealer\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Dealer::Dealer ()"

.PP
Default constructor\&. Creates a default deck of 52 cards\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016 
.RE
.PP

.PP
Definition at line 8 of file Dealer\&.cpp\&.
.PP
References createDeck(), and shuffleDeck()\&.
.PP
.nf
9 {
10     deck = createDeck();
11     shuffleDeck();
12     TableID = -1;
13     ID = nextID;
14     nextID++;
15 }
.fi
.SS "Dealer::Dealer (unsigned int ID)"

.PP
Constructor\&. Creates a default deck of 52 cards\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIID\fP The identifier\&. 
.RE
.PP

.PP
Definition at line 17 of file Dealer\&.cpp\&.
.PP
References createDeck(), and shuffleDeck()\&.
.PP
.nf
17                               {
18     deck = createDeck();
19     shuffleDeck();
20     TableID = -1;
21     this->ID = ID;
22 }
.fi
.SS "Dealer::Dealer (vector< \fBCard\fP > newDeck)"

.PP
Constructor\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fInewDeck\fP The new deck\&. 
.RE
.PP

.PP
Definition at line 24 of file Dealer\&.cpp\&.
.PP
References shuffleDeck()\&.
.PP
.nf
24                                    {
25     TableID = -1;
26     this->deck = newDeck;
27     shuffleDeck();
28     ID = nextID;
29     nextID++;
30 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void Dealer::addCardsToDeck (vector< \fBCard\fP > & cards)"

.PP
Adds cards to the deck\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIcards\fP Vector of cards to be added to the deck\&. 
.RE
.PP

.PP
Definition at line 84 of file Dealer\&.cpp\&.
.PP
References shuffleDeck()\&.
.PP
.nf
85 {
86     deck\&.insert(deck\&.end(), cards\&.begin(), cards\&.end());
87     shuffleDeck();
88 }
.fi
.SS "void Dealer::clearDiscardedDeck ()"

.PP
Clears the discarded deck\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016 
.RE
.PP

.PP
Definition at line 100 of file Dealer\&.cpp\&.
.PP
.nf
101 {
102     deckDiscarded\&.clear();
103 }
.fi
.SS "void Dealer::clearHand ()"

.PP
Clears the hand\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016 
.RE
.PP

.PP
Definition at line 105 of file Dealer\&.cpp\&.
.PP
References setHandScore()\&.
.PP
.nf
106 {
107     hand\&.clear();
108     this->setHandScore();
109 }
.fi
.SS "\fBCard\fP Dealer::discard (vector< \fBPlayer\fP *> players)"

.PP
Removes a \fBCard\fP from the end of the deck and returns it\&. Additionaly pushes the \fBCard\fP to the deckDiscarded vector\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIplayers\fP A vector of pointers to the players\&.
.RE
.PP
\fBReturns:\fP
.RS 4
The discarded \fBCard\fP\&. 
.RE
.PP

.PP
Definition at line 36 of file Dealer\&.cpp\&.
.PP
Referenced by hit(), Bot0::play(), Bot1::play(), Bot2::play(), and Human::play()\&.
.PP
.nf
36                                              {
37     Card nextCard;
38     nextCard = deck\&.back();
39     deckDiscarded\&.push_back(nextCard);
40     deck\&.pop_back();
41     for (size_t i = 0; i < players\&.size(); i++) {
42         players\&.at(i)->addCount(nextCard);
43     }
44     return nextCard;
45 }
.fi
.SS "vector< \fBCard\fP > & Dealer::getDeck ()"

.PP
Gets the deck\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The deck\&. 
.RE
.PP

.PP
Definition at line 74 of file Dealer\&.cpp\&.
.PP
.nf
75 {
76     return deck;
77 }
.fi
.SS "vector< \fBCard\fP > & Dealer::getDiscardedDeck ()"

.PP
Gets discarded deck\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The discarded deck\&. 
.RE
.PP

.PP
Definition at line 90 of file Dealer\&.cpp\&.
.PP
.nf
91 {
92     return deckDiscarded;
93 }
.fi
.SS "vector< \fBCard\fP > & Dealer::getHand ()"

.PP
Gets the hand\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The hand\&. 
.RE
.PP

.PP
Definition at line 79 of file Dealer\&.cpp\&.
.PP
Referenced by Bot1::play(), and Bot1::split()\&.
.PP
.nf
80 {
81     return hand;
82 }
.fi
.SS "unsigned int Dealer::getHandScore () const"

.PP
Gets handscore\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The handscore\&. 
.RE
.PP

.PP
Definition at line 95 of file Dealer\&.cpp\&.
.PP
Referenced by Bot1::surrender(), and Bot2::surrender()\&.
.PP
.nf
96 {
97     return handScore;
98 }
.fi
.SS "unsigned int Dealer::getID () const"

.PP
Gets the \fBDealer\fP's identifier\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The identifier\&. 
.RE
.PP

.PP
Definition at line 131 of file Dealer\&.cpp\&.
.PP
Referenced by Casino::addDealerToCasino(), DealerAlreadyExistException::DealerAlreadyExistException(), DealerIsOnTableAlreadyException::DealerIsOnTableAlreadyException(), DealerNotExistException::DealerNotExistException(), DealerStillOnTableException::DealerStillOnTableException(), and Casino::removeDealerFromCasino()\&.
.PP
.nf
131                                  {
132     return this->ID;
133 }
.fi
.SS "static unsigned int Dealer::getNextID ()\fC [static]\fP"

.PP
Gets the next identifier\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The next identifier\&. 
.RE
.PP

.PP
Definition at line 147 of file Dealer\&.cpp\&.
.PP
Referenced by Casino::saveDealersFile()\&.
.PP
.nf
147                                {
148     return nextID;
149 }
.fi
.SS "int Dealer::getTableOn () const"

.PP
Gets identifier of the table that the dealer is on\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The table that the dealer is on\&. -1 if \fBDealer\fP is not on any table\&. 
.RE
.PP

.PP
Definition at line 139 of file Dealer\&.cpp\&.
.PP
Referenced by Casino::create(), and DealerStillOnTableException::DealerStillOnTableException()\&.
.PP
.nf
139                              {
140     return this->TableID;
141 }
.fi
.SS "bool Dealer::hasCardInHand (\fBCard\fP & c)"

.PP
Query if dealer has card in hand\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIc\fP The \fBCard\fP to process\&.
.RE
.PP
\fBReturns:\fP
.RS 4
True if the card is in the dealer's hand, false if not\&. 
.RE
.PP

.PP
Definition at line 111 of file Dealer\&.cpp\&.
.PP
.nf
112 {
113     for (size_t i = 0; i < hand\&.size(); i++) {
114         if (hand\&.at(i) == c) {
115             return true;
116         }
117     }
118     return false;
119 }
.fi
.SS "\fBCard\fP & Dealer::hit (vector< \fBPlayer\fP *> players)"

.PP
Places a \fBCard\fP in the \fBDealer\fP hand and updates the handscore\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIplayers\fP A vector of pointers to the players\&.
.RE
.PP
\fBReturns:\fP
.RS 4
A reference to the dealer's new \fBCard\fP\&. 
.RE
.PP

.PP
Definition at line 47 of file Dealer\&.cpp\&.
.PP
References discard(), and setHandScore()\&.
.PP
Referenced by play()\&.
.PP
.nf
47                                           {
48     hand\&.push_back(discard(players));
49     setHandScore();
50     return hand\&.back();
51 }
.fi
.SS "string Dealer::play (\fBTable\fP & table)"

.PP
Calculates dealer next move\&. \fBDealer\fP will stand if handscore >= 17\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fItable\fP The table that the dealer is on\&.
.RE
.PP
\fBReturns:\fP
.RS 4
A string: 'hit' or 'stand' 
.RE
.PP

.PP
Definition at line 122 of file Dealer\&.cpp\&.
.PP
References Table::getPlayers(), and hit()\&.
.PP
.nf
123 {
124     if (handScore < 17) {
125         hit(table\&.getPlayers());
126         return "hit"; //means hit
127     }
128     else return "stand"; //means stand
129 }
.fi
.SS "unsigned int Dealer::setHandScore ()"

.PP
Updates dealer's handscore\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBReturns:\fP
.RS 4
The updated handscore\&. 
.RE
.PP

.PP
Definition at line 57 of file Dealer\&.cpp\&.
.PP
Referenced by clearHand(), and hit()\&.
.PP
.nf
57                                  {
58     this->handScore = 0;
59     for (size_t i = 0; i < hand\&.size(); i++)
60     {
61         this->handScore += hand\&.at(i)\&.score;
62     }
63     if (handScore > 21) {
64         for (size_t i = 0; i < hand\&.size(); i++) {
65             if (hand\&.at(i)\&.score == 11 && handScore > 21) {
66                 this->handScore -= 10;
67             }
68         }
69     }
70     return this->handScore;
71 }
.fi
.SS "static void Dealer::setNextID (unsigned int ID)\fC [static]\fP"

.PP
Sets next identifier\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fIID\fP The identifier\&. 
.RE
.PP

.PP
Definition at line 143 of file Dealer\&.cpp\&.
.PP
Referenced by Casino::readDealersFile()\&.
.PP
.nf
143                                       {
144     nextID = ID;
145 }
.fi
.SS "void Dealer::setTable (int tableID)"

.PP
Associates a table to this dealer\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016
.RE
.PP
\fBParameters:\fP
.RS 4
\fItableID\fP Identifier for the table\&. 
.RE
.PP

.PP
Definition at line 135 of file Dealer\&.cpp\&.
.PP
Referenced by Table::setDealer(), and Table::Table()\&.
.PP
.nf
135                                  {
136     this->TableID = tableID;
137 }
.fi
.SS "void Dealer::shuffleDeck ()"

.PP
Shuffle deck\&. 
.PP
\fBAuthor:\fP
.RS 4
M\&. Francisca Pauperio 
.RE
.PP
\fBDate:\fP
.RS 4
19/11/2016 
.RE
.PP

.PP
Definition at line 32 of file Dealer\&.cpp\&.
.PP
Referenced by addCardsToDeck(), and Dealer()\&.
.PP
.nf
32                          {
33     random_shuffle(deck\&.begin(), deck\&.end());
34 }
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for Aeda-Casino from the source code\&.


Dealer.cpp
#pragma once

#include <algorithm>
#include "Dealer.h"

unsigned int Dealer::nextID = 0;

Dealer::Dealer()
{
	deck = createDeck();
	shuffleDeck();
	TableID = -1;
	ID = nextID;
	nextID++;
}

Dealer::Dealer(unsigned int ID) {
	deck = createDeck();
	shuffleDeck();
	TableID = -1;
	this->ID = ID;
}

Dealer::Dealer(vector<Card> newDeck) {
	TableID = -1;
	this->deck = newDeck;
	shuffleDeck();
	ID = nextID;
	nextID++;
}

void Dealer::shuffleDeck() {
	random_shuffle(deck.begin(), deck.end());
}

Card Dealer::discard(vector<Player *> players) {
	Card nextCard;
	nextCard = deck.back();
	deckDiscarded.push_back(nextCard);
	deck.pop_back();
	for (size_t i = 0; i < players.size(); i++) {
		players.at(i)->addCount(nextCard);
	}
	return nextCard;
}

Card& Dealer::hit(vector<Player *> players) {
	hand.push_back(discard(players));
	setHandScore();
	return hand.back();
}
/*
void Dealer::stand() {
	nextPlayer++;
}*/

unsigned int Dealer::setHandScore(){
	this->handScore = 0;
	for (size_t i = 0; i < hand.size(); i++)
	{
		this->handScore += hand.at(i).score;
	}
	if (handScore > 21) {
		for (size_t i = 0; i < hand.size(); i++) {
			if (hand.at(i).score == 11 && handScore > 21) {
				this->handScore -= 10;
			}
		}
	}
	return this->handScore;
}


vector<Card>& Dealer::getDeck()
{
	return deck;
}

vector<Card>& Dealer::getHand()
{
	return hand;
}

void Dealer::addCardsToDeck(vector<Card>& cards)
{
	deck.insert(deck.end(), cards.begin(), cards.end());
	shuffleDeck();
}

vector<Card>& Dealer::getDiscardedDeck()
{
	return deckDiscarded;
}

unsigned int Dealer::getHandScore() const
{
	return handScore;
}

void Dealer::clearDiscardedDeck()
{
	deckDiscarded.clear();
}

void Dealer::clearHand()
{
	hand.clear();
	this->setHandScore();
}

bool Dealer::hasCardInHand(Card & c)
{
	for (size_t i = 0; i < hand.size(); i++) {
		if (hand.at(i) == c) {
			return true;
		}
	}
	return false;
}


string Dealer::play(Table &table)
{
	if (handScore < 17) {
		hit(table.getPlayers());
		return "hit"; //means hit
	}
	else return "stand"; //means stand
}

unsigned int Dealer::getID() const {
	return this->ID;
}

void Dealer::setTable(int tableID) {
	this->TableID = tableID;
}

int Dealer::getTableOn() const {
	return this->TableID;
}

void Dealer::setNextID(unsigned int ID) {
	nextID = ID;
}

unsigned int Dealer::getNextID() {
	return nextID;
}

DealerNotExist::DealerNotExist(Dealer *dealer) {
	this->dealerID = dealer->getID();
}

DealerIsOnTableAlready::DealerIsOnTableAlready(Dealer * dealer) {
	this->dealerID = dealer->getID();
}

DealerAlreadyExist::DealerAlreadyExist(Dealer * dealer) {
	this->dealerID = dealer->getID();
}

int DealerAlreadyExist::getID() const {
	return dealerID;
}

DealerStillOnTable::DealerStillOnTable(Dealer * dealer) {
	this->dealerID = dealer->getID();
	this->tableID = dealer->getTableOn();
}

unsigned int DealerStillOnTable::getTableID() const {
	return this->tableID;
}
